lib/internal/http2/core.js

```diff
@@ -2038,21 +2043,23 @@ const setTimeout = {
   enumerable: true,
   writable: true,
   value: function(msecs, callback) {
-    if (typeof msecs !== 'number') {
-      throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
-                                 'msecs',
-                                 'number');
-    }
+    // Type checking identical to timers.enroll()
+    msecs = validateTimerDuration(msecs);
+
     if (msecs === 0) {
-      unenroll(this);
+      clearTimeout(this[kTimeout]);
+
       if (callback !== undefined) {
         if (typeof callback !== 'function')
           throw new errors.TypeError('ERR_INVALID_CALLBACK');
         this.removeListener('timeout', callback);
       }
     } else {
-      enroll(this, msecs);
-      this[kUpdateTimer]();
+      this[kTimeout] = setUnrefTimeout(() => {
+        this._onTimeout();
+      }, msecs);
+      if (this[kSession]) this[kSession][kUpdateTimer]();
```

Comment:
Seems messy but it doesn't seem like `[kUpdateTimer]` should create new timers? 😕

Reply:
Maybe this is actually a bug? @jasnell should `HTTP2Stream#setTimeout()` create a timeout on the session, if there is a session?

Reply:
@Fishrock123 that line is to differentiate between HTTP2Stream and HTTP2Session, as this code is attached to both prototypes.

Reply:
Understood, but why refresh the session timer from the stream with it's existing timeout? Is that really intended from here without updating the session's timeout duration to a potential new duration?

Reply:
The session timeout needs to be refreshed every time any of the associated streams moves data, as the socket is there.